agent-instrumentation/src/jvmMain/kotlin/com/epam/drill/agent/instrument/servers/UndertowTransformerObject.kt

```diff
@@ -0,0 +1,87 @@
+import javassist.CtBehavior
+import javassist.CtClass
+import mu.KotlinLogging
+
+/**
+ * Transformer for simple Undertow-based web servers
+ *
+ * Tested with:
+ *      io.undertow:undertow-core:2.0.29.Final
+ */
+abstract class UndertowTransformerObject(
+    protected val headersRetriever: HeadersRetriever
+) : HeadersProcessor, AbstractTransformerObject() {
+
+    override val logger = KotlinLogging.logger {}
+
+    override fun permit(className: String?, superName: String?, interfaces: Array<String?>): Boolean =
+        "io/undertow/server/Connectors" == className
+
+    override fun transform(className: String, ctClass: CtClass) {
+        val adminHeader = headersRetriever.adminAddressHeader()
+        val adminUrl = headersRetriever.adminAddressValue()
+        val agentIdHeader = headersRetriever.agentIdHeader()
+        val agentIdValue = headersRetriever.agentIdHeaderValue()
+
+        logger.info { "transform: Starting UndertowTransformer with admin host $adminUrl..." }
+        val method = ctClass.getMethod("executeRootHandler", "(Lio/undertow/server/HttpHandler;Lio/undertow/server/HttpServerExchange;)V")
+
+        method.insertCatching(
+            CtBehavior::insertBefore,
+            """
+            if ($1 instanceof io.undertow.server.HttpHandler && $2 instanceof io.undertow.server.HttpServerExchange) {
+                io.undertow.util.HeaderMap responseHeaders = (io.undertow.util.HeaderMap) $2.getResponseHeaders();
+                if (!"$adminUrl".equals(responseHeaders.get("$adminHeader"))) {
+                    responseHeaders.add(io.undertow.util.HttpString.tryFromString("$adminHeader"), "$adminUrl");
+                    responseHeaders.add(io.undertow.util.HttpString.tryFromString("$agentIdHeader"), "$agentIdValue");
+                }
+        
+                io.undertow.util.HeaderMap requestHeaders = (io.undertow.util.HeaderMap) $2.getRequestHeaders();
+                java.util.Iterator/*io.undertow.util.HttpString>*/ headerNames = requestHeaders.getHeaderNames().iterator();
+                java.util.Map/*<java.lang.String, java.lang.String>*/ allHeaders = new java.util.HashMap();
+                while (headerNames.hasNext()){
+                    java.lang.String headerName = (java.lang.String) headerNames.next().toString();
+                    try {
+                        java.lang.String header = requestHeaders.get(io.undertow.util.HttpString.tryFromString(headerName)).element();
+                        allHeaders.put(headerName, header);
+        
+                        if (headerName.startsWith("${HeadersProcessor.DRILL_HEADER_PREFIX}") && responseHeaders.get(io.undertow.util.HttpString.tryFromString(headerName)) == null) {
+                            responseHeaders.add(io.undertow.util.HttpString.tryFromString(headerName), header);
+                        }
+                    } catch (java.util.NoSuchElementException ex) {
+                        //NOOP   
```

Comment:
Utilize `hasNext` to prevent undertow from throwing 
https://github.com/undertow-io/undertow/blob/128b9677ce82a32ef901d0b40f0ae9f647fdb09c/core/src/main/java/io/undertow/util/HeaderMap.java#L619C28-L619C35

Reply:
reworked